tcms/rpc/tests/test_testplan.py

```diff
@@ -226,3 +227,116 @@ def test_ignores_existing_mappings(self):
+        self.plan_type = PlanTypeFactory(name='api testing')
+        self.params = {
+            'product': self.product.pk,
+            'product_version': self.version.pk,
+            'name': 'Testplan foobar',
+            'type': self.plan_type.pk,
+            'text': 'Testing TCMS',
+            'parent': None,
+        }
+
+    def verify_api_with_permission(self):
+        plan = self.rpc_client.TestPlan.create(self.params)
+        self.assertIsNotNone(plan)
+        self.assertEqual(self.params['product'], plan['product_id'])
+        self.assertEqual(self.params['product_version'], plan['product_version_id'])
+        self.assertEqual(self.params['name'], plan['name'])
+        self.assertEqual(self.params['type'], plan['type_id'])
+        self.assertEqual(self.params['text'], plan['text'])
+        self.assertEqual(self.params['parent'], plan['parent'])
+
+    def verify_api_without_permission(self):
+        with self.assertRaisesRegex(ProtocolError, '403 Forbidden'):
+            self.rpc_client.TestPlan.create(self.params)
+
+
+class TestListAttachments(APITestCase):
+    def _fixture_setup(self):
+        super()._fixture_setup()
+
+    def test_list_attachments_with_wrong_plan_id(self):
+        with self.assertRaisesRegex(XmlRPCFault, 'TestPlan matching query does not exist'):
+            self.rpc_client.TestPlan.list_attachments(-1)
+
+
+class TestListAttachmentsPermissions(APIPermissionsTestCase):
+    permission_label = 'attachments.view_attachment'
+
+    def _fixture_setup(self):
+        super()._fixture_setup()
+
+        self.plan = TestPlanFactory()
+
+    def verify_api_with_permission(self):
+        attachments = self.rpc_client.TestPlan.list_attachments(self.plan.pk)
+        self.assertEqual(0, len(attachments))
+
+    def verify_api_without_permission(self):
+        with self.assertRaisesRegexp(ProtocolError, '403 Forbidden'):
+            self.rpc_client.TestPlan.list_attachments(self.plan.pk)
+
+
+class TestAddAttachment(APITestCase):
+    def _fixture_setup(self):
+        super()._fixture_setup()
+
+        self.plan = TestPlanFactory()
+
+    def test_add_attachment(self):
+        self.rpc_client.TestPlan.add_attachment(self.plan.pk, 'attachment.txt', 'kiwitcms')
+        attachments = self.rpc_client.TestPlan.list_attachments(self.plan.pk)
+        self.assertEqual(1, len(attachments))
+
+        attachment = attachments[0]
+        self.assertEqual(self.api_user.pk, attachment['owner_pk'])
+        self.assertEqual(self.api_user.username, attachment['owner_username'])
+        file_url = attachment['url']
+        fn = file_url[file_url.rindex('/') + 1:]
+        self.assertRegexpMatches(fn, r'^attachment_(.*).txt$')
+
+
+class TestAddAttachmentPermissions(APIPermissionsTestCase):
+    permission_label = 'attachments.add_attachment'
+
+    def _fixture_setup(self):
+        super()._fixture_setup()
+
+        self.plan = TestPlanFactory()
+
+    def verify_api_with_permission(self):
```

Comment:
It seems only one permission is supported here therefore it would be difficult to do list check in this case. In addition the scenario is implemented in test_add_attachment case, I just leave it pass.

Reply:
Query the database directly when you want to read data & verify it. We test the API method but no need to use API for the verification part.

Reply:
updated